### PR TITLE
alias expressions

### PR DESCRIPTION
**Pull request by wiedld**
#### Goal:
- to enable date ranges, relative to an unknown date at the time of odata WRITE.
- e.g. today, last week, last month

#### Solution:
- use odata alias expressions. Store odata with the aliasIdentifier in place. `@lx_date_last_week`
- then later on, at the time of odata READ, tack the aliasExpression on to the end of the odata (defining that the alias means). `...&@lx_date_last_week=dateCol ge '2016-11-04'` or `...&@lx_alias_standard_color='green' `
- This completed odata will be sent through the parser and the mapper.
- parser's objective is to build an ast subtree, such that the mapper can recognize both the aliasIdentifier as well as the aliasExpression.

#### odata Spec: see Parameter Alias
http://docs.oasis-open.org/odata/odata/v4.0/errata03/os/complete/part2-url-conventions/odata-v4.0-errata03-os-part2-url-conventions-complete.html#_Toc453752367

#### TODO in the Mapper:
- recognize the aliasIdentifier
- recognize the aliasExpression
- have the visitor be able to replace the aliasIdentifier with the aliasExpression
- since the aliasExpression will be at the end of the odata string, will need to do something different: 
(1) the mapper's visitor is already designed to do a single pass for transform + stringify.
(2) we can start making it like an actual compiler, with a series of IR transformations.
(3) make another visitor, to run earlier. Runs in 2 steps.
  -  1st run to find all the aliasExpr and build a hashmap (symbol table) and delete the aliasExpr nodes. 
  -  2nd run to replace all the aliasIdentifiers with the aliasExpr.

#### NOTE:
- [x] pegjs does build without errors
- [x] tests are written and passing.

### CODE REVIEW

**Comment by pythonesque** [comment]
The caveat doesn't seem thaaat bad.  We can probably require the @-expressions to be parsed prior to the normal parse tree.

**Comment by pythonesque** [comment]
👍

File: src/odata.pegjs [hunk 1]
```diff
@@ -296,6 +296,15 @@ orderbyList                 = i:(id:identifierPath ord:(WSP ("asc"/"desc"))? {
 select                      =   "$select=" list:selectList { return { "$select":list }; }
                             /   "$select=" .* { return {"error": 'invalid $select parameter'}; }
 
+// denote when an identifier is an alias. So the mapper can search for the matching aliasExpr
+aliasIdentifier            = "@" a:lambdaVar ":" b:lambdaVar {
```
Comment (wiedld) [inline src/odata.pegjs]: Note: I tried to make this have the path identifierPart "." identifierPart "." IdentifierPart etc....

However, it caused ambiguity in peg. This was a cleaner method.

Comment (tone81) [inline src/odata.pegjs]: Boo to Peg.

**Comment by pythonesque** [comment]
👍

**Comment by DannyDouglass** [comment]
💥 nice @wiedld!

**Pull request merged**